### PR TITLE
Add Go Client Generation

### PR DESCRIPTION
**Pull request by DomBlack**
This commit adds Go Client generation to the Encore CLI.

This client supports all types of API endpoints, and provides a couple of options allowing users to customize how they interact with it.

### CODE REVIEW

File: cli/cmd/encore/gen.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ Use '--env=local' to generate it based on your local development version of the
 
 Supported language codes are:
   typescript: A TypeScript-client using the in-browser Fetch API
+  go: A Go-Lang client"
```
Comment (eandre) [inline cli/cmd/encore/gen.go:39]: A `Go client`

Comment (eandre) [inline cli/cmd/encore/gen.go:39]:
```suggestion
  go: A Go client using net/http"
```

File: cli/cmd/encore/gen.go [hunk 2]
```diff
@@ -77,7 +78,7 @@ Supported language codes are:
 
 	genCmd.AddCommand(genClientCmd)
 	genClientCmd.Flags().StringVarP(&output, "output", "o", "", "The filename to write the generated client code to")
-	genClientCmd.Flags().StringVarP(&lang, "lang", "l", "", "The language to generate code for (only \"typescript\" is supported for now)")
+	genClientCmd.Flags().StringVarP(&lang, "lang", "l", "", "The language to generate code for (\"typescript\" or \"go\" are supported)")
```
Comment (eandre) [inline cli/cmd/encore/gen.go:81]:
```suggestion
	genClientCmd.Flags().StringVarP(&lang, "lang", "l", "", "The language to generate code for (\"typescript\" and \"go\" are supported)")
```

File: cli/internal/codegen/client.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ type Lang string
 const (
 	Unknown    Lang = ""
 	TypeScript Lang = "typescript"
+	GoLang     Lang = "go"
```
Comment (eandre) [inline cli/internal/codegen/client.go:22]:
```suggestion
	Go     Lang = "go"
```

File: cli/internal/codegen/client.go [hunk 4]
```diff
@@ -34,6 +35,8 @@ func Detect(path string) (lang Lang, ok bool) {
 	switch suffix {
 	case ".ts":
 		return TypeScript, true
+	case ".go":
+		return GoLang, true
```
Comment (eandre) [inline cli/internal/codegen/client.go:39]:
```suggestion
		return Go, true
```

File: cli/internal/codegen/client.go [hunk 5]
```diff
@@ -51,6 +54,8 @@ func Client(lang Lang, appSlug string, md *meta.Data) (code []byte, err error) {
 	switch lang {
 	case TypeScript:
 		gen = &ts{}
+	case GoLang:
```
Comment (eandre) [inline cli/internal/codegen/client.go:57]:
```suggestion
	case Go:
```

File: cli/internal/codegen/golang.go [hunk 6]
```diff
@@ -0,0 +1,821 @@
...
+
+		fieldDef = append(fieldDef,
+			Id(name).Id(fmt.Sprintf("%sClient", name)),
+		)
+
+		fieldInit[Id(name)] = Op("&").Id(fmt.Sprintf("%sClient", strings.ToLower(name))).Values(Id("base"))
+	}
+
+	// The client struct
+	file.Commentf("Client is a generated client for the Encore application %s.", appSlug)
```
Comment (eandre) [inline cli/internal/codegen/golang.go:76]:
```suggestion
	file.Commentf("Client is an API client for the %s Encore application.", appSlug)
```

File: cli/internal/codegen/golang.go [hunk 7]
```diff
@@ -0,0 +1,821 @@
...
+		Type().Id("Client").Struct(fieldDef...),
+		Line(),
+		Line(),
+	)
+
+	file.Comment("BaseURL is what URL your encore application is being hosted at")
+	file.Type().Id("BaseURL").String()
+	file.Line()
+
+	file.Const().Id("LocalDevelopment").Id("BaseURL").Op("=").Lit("http://localhost:4000")
```
Comment (eandre) [inline cli/internal/codegen/golang.go:87]:
```suggestion
	file.Const().Id("Local").Id("BaseURL").Op("=").Lit("http://localhost:4000")
```

File: cli/internal/codegen/golang.go [hunk 8]
```diff
@@ -0,0 +1,821 @@
...
+
+	// The client struct
+	file.Commentf("Client is a generated client for the Encore application %s.", appSlug)
+	file.Add(
+		Type().Id("Client").Struct(fieldDef...),
+		Line(),
+		Line(),
+	)
+
+	file.Comment("BaseURL is what URL your encore application is being hosted at")
```
Comment (eandre) [inline cli/internal/codegen/golang.go:83]:
```suggestion
	file.Comment("BaseURL is the base URL for calling the Encore application's API.")
```

File: cli/internal/codegen/golang.go [hunk 9]
```diff
@@ -0,0 +1,821 @@
...
+	file.Type().Id("BaseURL").String()
+	file.Line()
+
+	file.Const().Id("LocalDevelopment").Id("BaseURL").Op("=").Lit("http://localhost:4000")
+	file.Line()
+
+	file.Comment("Environment can be used to specify an environment by name, rather than using LocalDevelopment.")
+	file.Comment("")
+	file.Comment("This function doesn't check if the environment exists first, so it can cause runtime errors if")
+	file.Comment("you pass in an invalid environment name")
```
Comment (eandre) [inline cli/internal/codegen/golang.go:93]: I think the comment about runtime errors is a bit verbose and unnecessary:
```suggestion
	file.Comment("Environment returns a BaseURL for calling the cloud environment with the given name.")
```

File: cli/internal/codegen/golang.go [hunk 10]
```diff
@@ -0,0 +1,821 @@
...
+	file.Add(
+		Type().Id("Option").Op("=").
+			Func().Params(Id("client").Op("*").Id("baseClient")).Error(),
+		Line(),
+		Line(),
+	)
+
+	// New Function
+	file.Comment("New returns a Client for calling the public and authenticated APIs of your Encore application.")
+	file.Comment("You can customize the behaviour of the client using the given Option functions, such as WithHttpClient or WithAuthToken.")
```
Comment (eandre) [inline cli/internal/codegen/golang.go:117]:
```suggestion
	file.Comment("You can customize the behaviour of the client using the given Option functions, such as WithHTTPClient or WithAuthToken.")
```

**Review by ekerfelt — APPROVED** [review]

**Pull request merged**